lib/network/modules/LayoutEngine.js

```diff
@@ -1684,55 +1664,96 @@ class LayoutEngine {
     return findParent(parents, childB);
   }
 
+
   /**
-   * Abstract the getting of the position so we won't have to repeat the check for direction all the time
-   * @param {vis.Node} node
-   * @param {{x: Number, y: Number}} position
-   * @param {Number} level
-   * @param {boolean} [doNotUpdate=false]
+   * Set the strategy pattern for handling the coordinates given the current direction.
+   *
+   * The individual instances contain all the operations and data specific to a layout direction.
+   *
    * @private
    */
-  _setPositionForHierarchy(node, position, level, doNotUpdate = false) {
-    //console.log('_setPositionForHierarchy',node.id, position)
-    if (doNotUpdate !== true) {
-      this.hierarchical.addToOrdering(node, level);
-    }
+  setDirectionStrategy() {
+    var isVertical = (this.options.hierarchical.direction === 'UD'
+                   || this.options.hierarchical.direction === 'DU');
+
+    if(isVertical) {
+      this.direction = {
+        // Type to use to translate dynamic curves to, in the case of hierarchical layout.
+        // Dynamic curves do not work for there.
+        type: 'vertical',
+
+        getPosition: (node) => {
+          return node.x;
+        },
+
+        setPosition: (node, position, level = undefined) => {
+          if (level !== undefined) {
+            this.hierarchical.addToOrdering(node, level);
+          }
+          node.x = position;
+        },
+
+        // get the width of a tree
+        getTreeSize: (index) => {
+          let res = this.hierarchical.getTreeSize(this.body.nodes, index);
+          return {min: res.min_x, max: res.max_x};
+        },
+
+        // Test on 'undefined' takes care of divergent behaviour in chrome
+        sort: (a, b) => {
+          if (a.x === undefined || b.x === undefined) return 0;   // THIS HAPPENS
+          return a.x - b.x;
+        },
 
-    if(this._isVertical()) {
-      node.x = position;
-    }
-    else {
-      node.y = position;
-    }
-  }
+        // Fix the 'other' coordinate for the direction 
+        fix: (node, level) => {
+          node.y = this.options.hierarchical.levelSeparation * level;
+          node.options.fixed.y = true;
+        },
+
+        shift: (nodeId, diff) => {
+          this.body.nodes[nodeId].x += diff;
+        },
+      };
+    } else {
+      // See the vertical instance for comments.
```

Comment:
An alternative is to use jsdoc `@link` to link to the documentation

Reply:
I don't see how I can make this work here; I've kept `direction` intentionally light-weight, so it's not really a separate class (or otherwise) that can be referred to (I think).

Reply:
Ok